docs/docs/schema-customization.md

```diff
@@ -0,0 +1,591 @@
+interface, which describes the set of fields common to node objects created by
+source and transformer plugins (`id`, `parent`, `children`, as well as a couple
+of `internal` fields like `type`). In GraphQL Schema Definition Language (SDL),
+it looks like this:
+
+```graphql
+interface Node {
+  id: ID!
+  parent: Node!
+  children: [Node!]!
+  internal: Internal!
+}
+
+type Internal {
+  type: String!
+}
+```
+
+Types created by source and transformer plugins implement this interface. For
+example, the node type created by `gatsby-transformer-json` for `authors.json`
+will be represented in the GraphQL schema as:
+
+```graphql
+type AuthorJson implements Node {
+  id: ID!
+  parent: Node!
+  children: [Node!]!
+  internal: Internal!
+  name: String
+  firstName: String
+  email: String
+  joinedAt: Date
+}
+```
+
+## Automatic type inference
+
+It's maybe obvious but nevertheless important to note, that the data in
+`author.json` does not provide type information of the Author fields by itself.
+In order to translate the data shape into GraphQL type definitions, Gatsby has
+to inspect the contents of every field and check its type. In many cases this
+works very well and it is still the default mechanism for creating a GraphQL
+schema.
+
+There are however two problems with this approach: (1) it is quite
+time-consuming and therefore does not scale very well and (2) if the values on a
+field are of different types Gatsby cannot decide which one is the correct one.
+A consequence of this is that if your data sources change, type inference could
+suddenly fail.
+
+Both problems can be solved by providing explicit type definitions for Gatsby's
+GraphQL schema.
+
+## Creating type definitions
+
+Let's take the latter case first. Assume a new author joins the team, but in the
+new author entry there is a typo on the `joinedAt` field: "201-04-02" which is
+not a valid Date.
+
+```diff:title=src/data/author.json
++  {
++    "name": "Doe",
++    "firstName": "John",
++    "email": "john@example.com",
++    "joinedAt": "201-04-02"
++  }
+]
+```
+
+This will confuse Gatsby's type inference since the `joinedAt`
+field will now have both Date and String values.
+
+### Fixing field types
+
+To ensure that the field will always be of Date type, you can provide explicit
+type definitions to Gatsby with the `createTypes` action. It accepts type
+definitions in GraphQL Schema Definition Language:
+
+```js:title=gatsby-node.js
+exports.sourceNodes = ({ actions }) => {
+  const { createTypes } = actions
+  const typeDefs = `
+    type AuthorJson implements Node {
+      joinedAt: Date
+    }
+  `
+  createTypes(typeDefs)
+}
+```
+
+Note that the rest of the fields (`name`, `firstName` etc.) don't have to be
+provided, they will still be handled by Gatsby's type inference.
+
+> Although the `createTypes` action is passed to all `gatsby-node` APIs,
+> it has to be called before schema generation. We recommend to use the
+> `sourceNodes` API.
```

Comment:
```suggestion
> [`sourceNodes` API](https://gatsby.dev/api).
```

Reply:
Thanks! Let's directly link to `sourceNodes` docs though?

Reply:
I don't know if we can link to the section with the URL shortener